pdns/ws-auth.cc

```diff
@@ -2255,6 +2255,14 @@ static void apiServerCacheFlush(HttpRequest* req, HttpResponse* resp) {
 
   DNSName canon = apiNameToDNSName(req->getvars["domain"]);
 
+  SOAData sd;
+  UeberBackend B;
+  if (g_zoneCache.isEnabled() && B.getSOAUncached(canon, sd)) {
+    // zone exists (uncached), add/update it in the zone cache.
+    // Handle this first, to avoid concurrent queries re-populating the other caches.
+    g_zoneCache.add(canon, sd.domain_id);
+  }
+
   uint64_t count = purgeAuthCachesExact(canon);
```

Comment:
the `Exact` purge is not actually what the 'new zone' situation wants, but I don't immediately know how to fix that cleanly. I wonder if the API should just grow a suffix purge?

Reply:
For (some) consistency with `pdns_control`, maybe the API should append '$' and call `purgeCaches` instead (giving us a "subtree" purge). This is also what rfc2136handler.cc does.

Not sure if one should also clear DNSSECKeeper caches, as pdns_control does today.

Reply:
> For (some) consistency with `pdns_control`, maybe the API should append '$' and call `purgeCaches` instead (giving us a "subtree" purge). This is also what rfc2136handler.cc does.

This, or add a flag, not sure what is cleaner. Your proposal changes the behaviour - but we could argue it improves it.

> Not sure if one should also clear DNSSECKeeper caches, as pdns_control does today.

I'm thinking yes.

Reply:
I've tried to improve the cache cleaning bits so there's now a `purgeAuthCacheForZone` function that should do the right thing in all of these cases. I am not super convinced it actually covers everything, but thats what I can see from existing usage of separate cache cleaning functions. API now also uses `purgeAuthCacheForZone`, resulting in a zone cache, dnsseckeeper cache, query, packet cache flush.